torch/_higher_order_ops/triton_kernel_wrap.py

```diff
@@ -60,88 +62,176 @@ def reset_table(self) -> None:
+    debug = None
+    target = None
+
+    # Build kernel signature -- doesn't include constexpr arguments.
+    signature = {
+        i: kernel._type_of(kernel._key_of(arg))
+        for i, arg in enumerate(args)
+        if i not in kernel.constexprs
     }
 
-    def __init__(self, infos) -> None:
-        super().__init__()
-        self.infos = infos
-        self.read_depth = 0
-        self.in_store = False
+    @dataclasses.dataclass(frozen=True)
+    class TensorParam:
+        idx: int
+
+    @dataclasses.dataclass(frozen=True)
+    class NonTensorParam:
+        source: Any
+        pass
+
+    @dataclasses.dataclass(frozen=True)
+    class Intermediate:
+        idx: int
+
+    mappings: Dict[Any, Union[TensorParam, NonTensorParam, Intermediate]] = dict()
+    next_intermediate = 0
+
+    def convert(arg):
+        if isinstance(arg, triton._C.libtriton.triton.ir.block_argument):
+            if arg not in mappings:
+                mappings[arg] = NonTensorParam(arg)
+            return mappings[arg]
+        if isinstance(arg, triton._C.libtriton.triton.ir.value):
+            if arg not in mappings:
+                nonlocal next_intermediate
+                mappings[arg] = Intermediate(next_intermediate)
+                next_intermediate += 1
+            return mappings[arg]
+        return arg
+
+    # Name of mutation op to mutated parameter indices
+    MUTATION_OPS = {"masked_store": [0]}
+
+    ops: Dict[Union[TensorParam, NonTensorParam, Intermediate], "Op"] = dict()
+    sinks: List["Op"] = []
+
+    @dataclasses.dataclass
+    class Op:
+        name: str
+        args: List[Any]
+        kwargs: Dict[str, Any]
+
+        @staticmethod
+        def create(name, args, kwargs, result):
+            op = Op(name, [convert(a) for a in args], kwargs)
+            ops[convert(result)] = op
+            if name in MUTATION_OPS:
+                sinks.append(op)
+
+    class MutationAnalysisWrapper:
+        def __init__(self, builder):
+            self.builder = builder
+
+        def __getattr__(self, name):
+            def generic_indirection(name, *args, **kwargs):
+                result = getattr(self.builder, name)(*args, **kwargs)
+
+                if name.startswith("create_"):
+                    Op.create(name[len("create_") :], args, kwargs, result)
+                elif name in {"get_int32"}:
+                    Op.create(name[len("get_") :], args, kwargs, result)
+                elif name in {"get_loc", "set_loc"}:
+                    pass
+                else:
+                    pass
+                return result
+
+            return functools.partial(generic_indirection, name)
+
+    original_add_entry_block = triton._C.libtriton.triton.ir.function.add_entry_block
+
+    def custom_add_entry_block(self):
+        result = self.original_add_entry_block()
+        for i, arg in enumerate(args):
+            if isinstance(arg, Tensor):
+                mappings[self.args(i)] = TensorParam(i)
```

Comment:
Here I believe your mapping is from a python wrapper object returned by one call to `block.args(i)`; then the actual IR value used throughout the triton code is a second python wrapper object created by the `args` call in `OriginalCodeGenerator`. 

It is also suspicious that you're using `block.args` whereas triton is using `function.args`, it's possible those are not the same even in C++.

Reply:
You might have misread something? self here is `mlir::triton::FuncOp`, so `self.args` is args function on the `ir.Function`, not `ir.Block`. Notice how this function is overwriting `triton._C.libtriton.triton.ir.function.add_entry_block`.

Reply:
I am intercepting `add_entry_block` because until this function executes, ir.Function does not have `args()` setup yet.

Reply:
Okay so in that case it must be that `function.args(i)` is returning a new wrapper object on the second call. 

I don't understand why you wouldn't just intercept `args` itself though?